feat: add bidirectional motion repeat with selection support

willbush opened the pull request:
Closes #12877
Fixes #8761

Related discussion: https://github.com/helix-editor/helix/discussions/8495

This continues @Aqaao's initial work [here](https://github.com/helix-editor/helix/issues/12092#issuecomment-2660711956).
Taking into consideration @the-mikedavis's [comment here](https://github.com/helix-editor/helix/issues/8761#issuecomment-1803145453).

- Add reverse motion repeat (Alt-Shift-.) to complement existing repeat (Alt-.)

  I'm completely open to suggestions for an alternative key binding. I
  personally use a heavily modified keymap to support Colemak-DHm. I'm planning
  to bind this to `'` / `"` (select register `&`) in my config.

- New commands:
  - `repeat_last_motion_reverse`
  - `extend_repeat_last_motion`
  - `extend_repeat_last_motion_reverse`

- Add extend variants for both forward/reverse motion repeat

  Why add extend variants of `repeat_last_motion`? This is motivated by
  @pascalkuthe's
  [comment](https://github.com/helix-editor/helix/issues/8761#issuecomment-1802795771):

  >This is the only consistent behaviour because select mode is not supposed to
  >be special. Instead it should just be a keymap like window mode/space
  >mode/goto mode,... You can already bind extend_next_char in normal mode and
  >it would be strange if that would extend the selection (similarly it would be
  >strange if you biud find_next_char in visual mode and itwould be similarly
  >strange if reperating that would extend the selection.

  The extend variants override the last `movement` (i.e Move / Extend) to always extend.

- Update documentation and tutor

- Note this also fixes some issue I run into when using find / till on `<ret>`
  Basically, it works as expected now. I haven't searched through existing
  issues to see if any can be closed yet.

## Possible improvements

I haven't looked into adding tests.

## Short demos

find / till `.` either starting in normal or select mode and switching:

https://github.com/user-attachments/assets/6f8f209f-5eb4-4ce8-8598-1e89ec2a2c4f

repeat motion and reverse repeat on next paragraph:

https://github.com/user-attachments/assets/7bc19f4d-034b-4f07-9374-ab9130f17be9

repeat motion and reverse on next function:

https://github.com/user-attachments/assets/d653206d-03ef-4930-a10b-089f0b9233a1

willbush commented on helix-view/src/editor.rs:
```diff
@@ -1236,19 +1243,44 @@ impl Editor {
     }
 
     pub fn repeat_last_motion(&mut self, count: usize) {
+        self.repeat_last_motion_impl(count, MotionMode::Normal, Some(Movement::Move));
```
On second thought. Perhaps non-extend repeat commands should not override previous movement to `Movement::Move`?

nik-rev commented:
I'm also using Colemak DH and Alt + Shift anything is literally impossible for me to press. Nice to see someone else using it!

I think A-S-. is the one that makes the most sense and one that feels consistent with other keybindings in the editor

willbush commented:
> I'm also using Colemak DH and Alt + Shift anything is literally impossible for me to press. Nice to see someone else using it!
> 
> I think A-S-. is the one that makes the most sense and one that feels consistent with other keybindings in the editor

Think @pascalkuthe uses it too! What keyboard you guys use? Something with qmk firmware I'm guessing. I use this little thing 
https://github.com/foostan/crkbd so yea I'm also not a fan of alt-S bindings either. :D

nik-rev commented:
> > I'm also using Colemak DH and Alt + Shift anything is literally impossible for me to press. Nice to see someone else using it!
> > I think A-S-. is the one that makes the most sense and one that feels consistent with other keybindings in the editor
> 
> Think @pascalkuthe uses it too! What keyboard you guys use? Something with qmk firmware I'm guessing. I use this little thing https://github.com/foostan/crkbd so yea I'm also not a fan of alt-S bindings either. :D

This is my layout

![image](https://github.com/user-attachments/assets/255fd26c-ccdb-4f01-a052-53d2d63b6bac)
![image](https://github.com/user-attachments/assets/01867733-5b27-48f9-b993-9750cd9f16c1)
![image](https://github.com/user-attachments/assets/4eee919b-ae5e-46e2-a033-a92de6f46da7)

I'm also using corne keyboard! Really comfortable. The numbers are in weird positions, [but I placed them there for a reason.](https://www.reddit.com/r/KeyboardLayouts/comments/j4vt1s/optimizing_the_number_row_essay_script/) :)

willbush commented:
> I was trying to find a solution for how I can possibly have a natural combination for Shift + Alt. Tried for a long time, but nothing feels good. Everything adjustments feels like a compromise to make that combination possible. So I've just decided to remap any Shift + Alt I come across.

Following is off-topic nerding out about keyboards for anyone reading this PR:

[Here's mine](https://github.com/willbush/system/blob/main/configs/keyboard-firmware/crkbd/keymap/keymap.c)

> The numbers are in weird positions

Ha love it! looks like the type of insane micro-optimization I would try.

One thing that stands out to me about your layout are the arrow keys on the primary layer. Personally, I put these on the raise layer (right mid thumb cluster button for me) in the same position as vim movement keys. Seems like that would be easier to reach and would also free up a lot of possibilities on your primary layer.

I have one deviation from Colemak-DH. I swap the `,` and `y` key. When learning the layout I realize the thing I hate most about the layout is typing the words "you're", "your", and "you". The word "you" et al. is one of the most common words in English and is insanely hard to type in the standard Colemak positions.

I tried the swap on: https://colemakmods.github.io/mod-dh/analyze.html and this is an improvement in score on Colemak-DH due to swapping the same-finger bigrams `e,` with the `ey` which occurs about half as often. Typing the word "you" et al. is still not amazing, but it's much easier than before. Another minor downside is words that end with `ly` don't finger roll as nicely on the top row.

Anyway, moving arrow keys off main layer opens up a lot of things you can do. You can move `,` back to the right. have a dedicated alt key etc.

Personally, I like having number row across all my main layers so I can hold the layer down with the opposite hand that's pressing the number key. I get away with this by having a dedicated layer for F keys (LOWER + v takes me to FN layer).

Interesting that you press space with your left hand. I've always wondered if that would be a better balance in typing across both hands!

willbush commented:
> This is my layout

Also funny to me that you're not the first person I've run into that uses this keyboard: https://github.com/meow-edit/meow/discussions/626#discussioncomment-10310662

Thought this would be a lot more rare.

willbush commented:
Reviewing this again, I realized it makes sense to be able to reverse the motion of `expand_selection` / `shrink_selection`. So I made a commit for that.

Here a breakdown of motions where reverse repeat last motion applies:

### applies:

```
goto_prev_paragraph
goto_next_paragraph

find_till_char
find_next_char
extend_till_char
extend_next_char
till_prev_char
find_prev_char
extend_till_prev_char
extend_prev_char

goto_next_diag
goto_prev_diag

goto_next_change
goto_prev_change

select_next_sibling
select_prev_sibling

move_parent_node_end
move_parent_node_start
extend_parent_node_end
extend_parent_node_start

goto_next_function
goto_prev_function
goto_next_class
goto_prev_class
goto_next_parameter
goto_prev_parameter
goto_next_comment
goto_prev_comment
goto_next_test
goto_prev_test
goto_next_entry
goto_prev_entry

expand_selection
shrink_selection
```
### does not apply

```
select_all_siblings
select_all_children

select_textobject_around
select_textobject_inner
```

willbush commented:
>select_textobject_around
>select_textobject_inner

Hmm just realized `map` "match around paragraph" + repeat last motion has directionality.. In this case I would expect revers repeat last motion to work.

nik-rev commented:
> > This is my layout
> 
> Also funny to me that you're not the first person I've run into that uses this keyboard: [meow-edit/meow#626 (comment)](https://github.com/meow-edit/meow/discussions/626#discussioncomment-10310662)
> 
> Thought this would be a lot more rare.

nice to know I'm not alone!

nik-rev commented:
> One thing that stands out to me about your layout are the arrow keys on the primary layer. Personally, I put these on the raise layer (right mid thumb cluster button for me) in the same position as vim movement keys. Seems like that would be easier to reach and would also free up a lot of possibilities on your primary layer.

Placing mine on the base layer allows me to move in insert mode with ease. Also the 4 arrow keys act as 4 free modifiers in normal / select mode

the-mikedavis commented on helix-view/src/editor.rs.orig:
Looks like this was checked in by accident?

the-mikedavis commented on runtime/tutor:
The tutor sections have a consistent height (# of lines between headers). For any new lines we add we should remove blank ones. If a section becomes too long and can't fit in that height we should split a section into twwo

the-mikedavis commented on helix-view/src/editor.rs:
```diff
@@ -441,6 +440,8 @@ pub fn get_terminal_provider() -> Option<TerminalConfig> {
     None
 }
 
+#[cfg(not(any(windows, target_arch = "wasm32")))]
```
Looks like this was moved when resolving a merge conflict - it should be right above `get_terminal_provider`

willbush commented:
@the-mikedavis Thanks for the review! I addressed issues you brought up, rebased on master, ran tests / integration test, and did some manual testing as a sanity check.

littleblack111 commented:
can we add an option to allow repeat of motion such as 10k 10j etc.
